### ISSUE TITLE
Support PROXY protocol

### ISSUE THREAD

**Issue by mitchellh** [eBay/fabio#97]
Hey there. Loving Fabio. :) Using it for some personal stuff.

I'm running Fabio behind an ELB and I noticed some odd behavior that I'm not sure is intended. If I setup a service to listen on "foo.com" but then make an ELB listen on port 1234, every HTTP client I've tried so far (easies with just curl) will set the `Host` header to `foo.com:1234`. This causes a 404 with Fabio.

I've worked around it by specifying the port in the Consul tag, but perhaps Fabio should strip the port?

**Comment by magiconair** [comment]
Hi @mitchellh, Thanks and that is nice to hear. I might see you in Amsterdam during Hashiconf 2016.

Matching on the full host header is indeed the intended behavior but maybe I'm missing something here. If it wouldn't do that then you couldn't run different services on `foo.com:81` and `foo.com:82` for example since they would both be routed to `foo.com`. What I've recently added though is dropping of the default port. See #90 

What is your use case?

**Comment by magiconair** [comment]
I've just merged #93 from @dkong which adds glob matching but only to the path. That should be simple to extend to the host as well.

**Comment by mitchellh** [comment]
Ah fair enough. I'm happy enough with the workaround. Glob matching to the host would be great but I don't need it (I can statically set the port).

Well, the issue is actually that I need to run a pure SSL => TCP listener on the ELB to allow websocket connections through, since HTTP on ELB won't allow it. However, I want to run the HTTP endpoints for 80/443 for normal web traffic so that I can get forwarded headers.

SO, the actual solution here would be if fabio supported the HAProxy proxy protocol that ELB also uses, then I can just run everything in TCP mode and avoid all of this. :)

**Comment by magiconair** [comment]
Why can't you just run 80/443 in TCP mode and terminate SSL on fabio? Sure, you'd lose the original src ip but you would lose that anyway if you're not going through some kind of protocol gateway. 

Also, how would fabio supporting PROXY help here? Assuming the originating request is coming from a regular HTTP client and ELB running in TCP mode you also wouldn't have an X-Forwarded-For header with the src ip or am I missing something? I'm not an expert on ELB though ...

**Comment by mitchellh** [comment]
hey @magiconair, thanks for the quick response!

I can't terminate SSL on Fabio because I'm using Amazon Certificate Manager which only allows certs to be tied to a few things (ELBs, Cloudfront, probably some others). They don't give you the certs directly. 

The proxy protocol is what the ELB uses to send the X-Forwarded-* equivalent information on TCP connections. See here: http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/enable-proxy-protocol.html

**Comment by magiconair** [comment]
I guess I could replace the http listener with the one from Armons repo: https://github.com/armon/go-proxyproto or at least use it as a basis. AFAIU, it just detects the `PROXY` line and then stores the addresses in the `Conn` struct. Doesn't look difficult.

**Comment by magiconair** [comment]
@mitchellh I'm still fighting with setting things up on AWS but from the look of it that should be all that's necessary. I'll test it once I have things running on AWS but if you want to give it a spin please feel free.

**Comment by magiconair** [comment]
Got it to work. I've added a simple log line to show the remote addr. (access logging is #80)

With HTTP LB:

```
Got request from 10.0.1.58:62422 for /echo
```

With TCP LB and patch for #97:

```
PROXY TCP4 82.161.239.242 192.168.9.213 63477 80.

Got request from 82.161.239.242:63477 for /echo
```

**Comment by mitchellh** [comment]
Hah yeah setting up/enabling PROXY protocol is a PITA. Use Terraform ;) (if you weren't already).

That looks great! Would it be possible to actually inject the X-Forwarded-* headers from the PROXY protocol? I'm asking for a lot more here but that'd be great. But this is already really wonderful. Thanks!

**Comment by magiconair** [comment]
I was. Even found https://github.com/hashicorp/terraform/pull/1749. Only had to figure out what to set as policy (`ProxyProtocol`) :)

That's what I'm currently getting:

```
GET /echo HTTP/1.1.
Host: fabio-example-elb-1332123959.eu-west-1.elb.amazonaws.com.
User-Agent: Go-http-client/1.1.
Connection: Upgrade.
Forwarded: for=82.161.239.242; proto=http; by=10.0.1.248.
Origin: http://localhost/.
Sec-Websocket-Key: t9mUKAGGLUFiEwpz7aCN5w==.
Sec-Websocket-Version: 13.
Upgrade: websocket.
X-Forwarded-Proto: http.
X-Real-Ip: 82.161.239.242.
```

The `X-Forwarded-For` header is added automatically by the HTTP reverse proxy but not for the ws proxy. Also, the `X-Forwarded-Proto` header should probably contain `ws` or `wss`. I'll fix it.

**Comment by magiconair** [comment]
I've fixed the header handling in #98 and merged the change for #97 to master. I've also added support for the `X-Forward-Port` header while I was at it. The result is:

```
GET /echo HTTP/1.1.
Host: fabio-example-elb-1332123959.eu-west-1.elb.amazonaws.com.
User-Agent: Go-http-client/1.1.
Connection: Upgrade.
Forwarded: for=82.161.239.242; proto=ws; by=10.0.1.248.
Origin: http://localhost/.
Sec-Websocket-Key: FJXs4MaHM29ID8+Yp3EnCg==.
Sec-Websocket-Version: 13.
Upgrade: websocket.
X-Forwarded-For: 82.161.239.242.
X-Forwarded-Port: 58686.
X-Forwarded-Proto: ws.
X-Real-Ip: 82.161.239.242.
```

Let me know whether that does the trick.

**Comment by mitchellh** [comment]
Looks awesome. I'll give it a try tonight or sometime (using Fabio for a for-fun project). Is the Docker container automatically updated? Using that as wel.

**Comment by magiconair** [comment]
No, but it might be a good time for a 1.1.3rc1. I'll spin one

**Comment by magiconair** [comment]
Done - both release and Docker container. Let me know how it goes.

**Comment by mitchellh** [comment]
Thanks will let you know tonight or tmrw night!

**Comment by magiconair** [comment]
There is an issue with Armons proxy listener. For some reason the request object no longer has the `tls.ConnectionState` set in `TLS` which means I can't set the protocol headers accordingly. I need to investigate this but for now consider `v1.1.3rc1` broken.

**Comment by magiconair** [comment]
I think I just used it in the wrong place. Testing now

**Comment by magiconair** [comment]
Should be fixed in 1.1.3rc2

**Comment by magiconair** [comment]
I've release 1.1.3 today and I am closing this ticket for now. Pls re-open if you have issues.
